---
title: "Clean-up the disease attributes of cMD sample metadata"
author: "Sehyun Oh"
date: "`r format(Sys.time(), '%B %d, %Y')`"
format:
    html:
        fontsize: 14px
        toc: true
        top-depth: 3
abstract: "Prepare U24 Supplement: AI/ML-ready"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, 
                      collapse = TRUE)
```

# Load Packages 
```{r}
library(gsheet)
library(dplyr)
library(RCurl) # for getURL
library(curatedMetagenomicData)
library(OmicsMLRepoData)
library(readr)
library(tidyr)
library(yaml)
library(googlesheets4)
library(tidyverse)
```

# Collect curations files 
Manual curation efforts for curatedMetagenomicData `sampleMetadata` are 
happening in a several GitHub repositories. We will first collect the most
up-to-date versions of them. Data inputs include: 

1. [Updated-disease-ontology.csv](https://github.com/waldronlab/curatedMetagenomicDataCuration/blob/master/inst/extdata/updated-disease-ontology.csv)
2. [Body-site-ontology.csv](https://github.com/waldronlab/curatedMetagenomicDataCuration/blob/master/inst/extdata/body-site-ontology.csv)
3. [Disease_ontology.yaml](https://github.com/Ethan-Loo/Bioinformatics_Data_Curation_Collaboration/blob/main/disease_ontology.yaml)
4. [sampleMetadata.rda](https://github.com/waldronlab/curatedMetagenomicData/raw/master/data/sampleMetadata.rda) 

## Import Data 

### curatedMetagenomicDataCuration

Disease ontology CSV
```{r}
file_url <- "https://raw.githubusercontent.com/waldronlab/curatedMetagenomicDataCuration/master/inst/extdata/updated-disease-ontology.csv"
download.file(file_url, "updated-disease-ontology.csv")
cmd_disease <- read.csv("updated-disease-ontology.csv")
head(cmd_disease, 3)
```

Completed cMD_bodysite, body-site-ontology.csv
```{r}
file_url <- "https://raw.githubusercontent.com/waldronlab/curatedMetagenomicDataCuration/master/inst/extdata/body-site-ontology.csv"
download.file(file_url, "body-site-ontology.csv")
cMD_bodysite <- read.csv("body-site-ontology.csv")
head(cMD_bodysite, 5)
```

### sample Metadata from cMD
```{r}
file_url <- "https://github.com/waldronlab/curatedMetagenomicData/raw/master/data/sampleMetadata.rda"
download.file(file_url, "sampleMetadata.rda")
load("sampleMetadata.rda")
```

### Bioinformatics_Data_Curation_Collaboration/disease_ontology.yaml

Note: This chunk needs to be fixed. Running into error. 
```{r}
file_url <- "https://github.com/Ethan-Loo/Bioinformatics_Data_Curation_Collaboration/blob/main/disease_ontology.yaml"
download.file(file_url, "disease_ontology.yaml")
# disease_ontology <- yaml::yaml.load_file("disease_ontology.yaml")
# head(disease_ontology, 3)
```

# Merging Curation Files

## Disease

Merge 'study_condition' and 'disease_subtype' columns. 
```{r}
unique_subtype <- unique(sampleMetadata$disease_subtype)
subtype_all <- c() # collect all the subtypes
for (i in seq_along(unique_subtype)) {
    subtype <- strsplit(unique_subtype[i], split = ";") # split the split semicolon-separated values
    subtype_all <- c(subtype_all, unlist(subtype))
}
    
## All the unique study_condition and disease_subtypes
unique_disease <- c(unique(sampleMetadata$study_condition), unique(subtype_all)) 
## Format it as same as cmd_disease table
unique_disease_tb <- data.frame(Original.Abbreviation.Name = sort(unique_disease), 
                                Full.Name = NA,
                                Ontology.link = NA)
```

Next, we should combine `unique_disease_tb` and `cmd_disease` tables 
without losing any information.

Vertical merge `unique_disease_tb` and `cmd_disease` tables 
```{r}
cmd_disease_unique_disease <- rbind(cmd_disease, unique_disease_tb)
```

Create ont_complete by including only unqiue values and filter rows where 
"ontology" col is not empty into 1 DF "ont_complete" 
```{r}
ont_complete <- cmd_disease_unique_disease %>% distinct(Original.Abbreviation.Name, .keep_all=TRUE) %>% filter(Ontology.link != "")
```

Create manaul_curation by including only unqiue values and filter rows where 
"ontology" col is empty (i.e., missing ontology link) into 1 DF "manaul_curation" 
```{r}
# Filter rows where manual curation is needed
manaul_curation <- cmd_disease_unique_disease %>% distinct(Original.Abbreviation.Name, .keep_all=TRUE) %>% filter(Ontology.link == "")

# Add a new column for "Revision Required" that explains observations that require clarification
manaul_curation <- manaul_curation %>% add_column(Revision_Required = "")
```

Temporal file for manual curation. Export the merged mapping table to Google Sheet 
for manual curation.
```{r}
(ss <- gs4_create("cMDOntology_NeedsCuration", sheets = manaul_curation))
```

Save manually curated mapping table. After the manual curation, update the target 
mapping table in GitHub. 
```{r}
manual_curation <- read_sheet("https://docs.google.com/spreadsheets/d/1Vai_5gfH3UYg3F5eyA27NytqfAZMN0805aE8MBSQstk/edit?usp=sharing")
```

Merge curated data with the cleaned DF 
```{r}
cMD_disease <- full_join(manual_curation, ont_complete, by = NULL)
```

Rename Full.Name column header to Full.Name.Curated
```{r}
cMD_disease <- rename(cMD_disease, "Full.Name" = "Full.Name.Curated")
```

Export to .CSV 
```{r}
write.csv(cMD_disease, "~/OmicsMLRepoData/curatedMetagenomicData/Disease_Curation.csv", row.names=FALSE)
```

Next Steps: Need to repeat steps, shown above, for age. Work with team to 
strategize ways to structure this data.  

## Age

Merge 'age', 'infant_age', and 'gestational_age' columns.


# Appendix 

Used the following chunk to review Disease_Curation Google sheet: 

library(curatedMetagenomicData)
getPMID <- function(key, value) {
    ind <- grep(value, sampleMetadata[,key])
    print(unique(sampleMetadata$PMID[ind]))
}

